Updated Rate-limiting library (#2)
slashid/ratelimit: README.md

- Multiple state storage options:

  - In-memory, for super-fast performance in single-process scenarios
  - Using a Redis server for shared ratelimiting between multiple processes, with multiple variants:

    - Redis module, for best performance (But requires some setup)
    - Lua script, for excellent performance with minimal setup
    - Regular GET/SET calls, if you absolutely want to keep custom code.

      This variant doesn't use atomic operations, therefore your target rate limits may ocasionally be exceeded due to concurrency issues.

# Rate-limiting parameters

There are many different algorithms (and implementations) out there, each one using different set of parameters, which can be quite confusing.

However they are usually equivalent, and it is easy to convert between them!

| Name                   | Description                                                                          | Example               | Equivalence                      |
| ---------------------- | ------------------------------------------------------------------------------------ | --------------------- | -------------------------------- |

> | `interval` or `period` | amount of time that should be elapsed between requests                               | a request every 10s   | `window/burst`, `1/rate`         |

Reviewer: ```suggestion
| `interval` or `period` | Amount of time between requests                               | a request every 10s   | `window/burst`, `1/rate`         |
```

| `rate` or `frequency`  | How often requests can be made                                                       | 6 requests per minute | `burst/window`, `1/interval`     |
| `burst`                | Maximum amount of requests that can be executed at once                              | 50 requests           | `window/interval`, `window*rate` |

> | `window`               | Amount of time to be considered when deciding if a request exceeds to the rate limit | 5-minute window       | `interval*burst`, `burst/rate`   |

Reviewer: ```suggestion
| `window`               | Amount of time to consider when deciding if a request exceeds the rate limit | 5-minute window       | `interval*burst`, `burst/rate`   |
```

In this implementation we use `interval` and `window`. Having all parameters (and the clock) in time units is convenient and allows a simpler implementation.

> # Comparisson of Rate Limiting Algorithms

Reviewer: ```suggestion
# Comparison of Rate Limiting Algorithms
```

There are many different algorithms to implement rate limiting, each with its pros and cons.

## No burst

The simplest algorithm ensures the required interval has elapsed since serving the last request.

This approach is very simple and efficient, but doesn't allow bursts at all, which is usually desirable in real-world scenarios. It is equivalent to having `burst=1` in other algorithms.

```python
class NoBurstRateLimiter:
    def __init__(self, interval: float):
        self.interval = interval  # seconds between requests
        self.ready_after = time.time()  # # timestamp when the next request will be allowed -- ready now

    def __call__(self) -> bool:
        now = time.time()
        if now < self.ready_after:
            # reject the request if `ready_after` hasn't been reached yet.
            return False
        else:
            # new requests will only be allowed after the interval has elapsed.
            self.ready_after = now + self.interval
            return True
```

## Fixed Window

On fixed window algorithm, time is divided in discrete intervals, like every full minute, hour or day.

> During that time window you can distribute the number of requests in any way you want: Evenly spread during the interval, all at once or something in-between. Either way, once exausted you will need to wait until the next window boundary (e.g., next midnight, next full hour or next full minute).

Reviewer: ```suggestion
During that time window you can distribute the number of requests in any way you want: evenly spread during the interval, all at once, or something in-between. Either way, once exhausted you will need to wait until the next window boundary (e.g., next midnight, next full hour or next full minute).
```

> This algorithm is very easy to reason about and simple to implement, but has its drawbacks:

Reviewer: ```suggestion
This algorithm is very easy to reason about and simple to implement, but has a major drawback:
```

> It goes immediately to fully burst-ready at the start of each window. On worst case it allows 2X the burst limit to be served in a short interval.

Reviewer: ```suggestion
it goes immediately to fully burst-ready at the start of each window. In the worst case it allows 2x the burst limit to be served in a short interval.
```

```python
class FixedWindowRateLimiter:
    def __init__(self, burst: int, window: float):
        self.burst = burst
        self.window = window
        self.count = 0  # Current window has served 0 requests
        self.current_window = time.time() // self.window  # infer current window sequence from timestamp

    def __call__(self) -> bool:
        current_window = time.time() // self.window # infer current window sequence from timestamp

        # reset the counter if we moved to a new window
        if current_window != self.current_window:
            self.count = 0
            self.current_window = current_window

        if self.count >= self.burst:
            # reject the request if the current window is exausted
            return False
        else:
            # update the request counter
            self.count+=1
            return True
```

## Sliding Window

This is conceptually similar to the fixed window algorithm, but we are now counting how many requests were executed in the last `window` seconds (instead of examining a fixed interval).

> This has some advantages over the fixed window algorithm: It avoids the 2X burst problem near the end of the window, and it doesn't recovers the full burst capacity at predetermined times.

Reviewer: ```suggestion
This has some advantages over the fixed window algorithm: it avoids the 2x burst problem near the end of the window, and it doesn't recover the full burst capacity at predetermined times.
```

However the implementation is considerably more complex and requires more memory:

- An exact solution now requires keeping track of all request timestamps inside the window

> - An efficient (but approximate) solution can be achieved by adding the counter of several smaller fixed windows (e.g., you can simulate the counter of a 1-minute sliding window by adding the last 60 1-second fixed windows - [Take a look at this Figma blog post for more info](https://www.figma.com/blog/an-alternative-approach-to-rate-limiting/#fixed-window-counters)

Reviewer: ```suggestion
- An efficient (but approximate) solution can be achieved by adding the counter of several smaller fixed windows (e.g., you can simulate the counter of a 1-minute sliding window by adding the last 60 1-second fixed windows - [take a look at this Figma blog post for more info](https://www.figma.com/blog/an-alternative-approach-to-rate-limiting/#fixed-window-counters).
```

> (TODO: Improve this) Another quirk of sliding windows is that capacity increases in the same pattern as the usage in the previous window. E.g., If you get burst of 10 requests, an idle interval and then another burst that exausts your rate limit, in the next window you will also get an increase of 10 in the burst capacity, then no changes during an interval and then full burst capacity. Instead it is preferable to increase it continuously every `window/burst` seconds (See Leaky Bucket below)

Reviewer: Still TODO?

> (TODO: Improve this) Another quirk of sliding windows is that capacity increases in the same pattern as the usage in the previous window. E.g., If you get burst of 10 requests, an idle interval and then another burst that exausts your rate limit, in the next window you will also get an increase of 10 in the burst capacity, then no changes during an interval and then full burst capacity. Instead it is preferable to increase it continuously every `window/burst` seconds (See Leaky Bucket below)

Reviewer: ```suggestion
(TODO: Improve this) Another quirk of sliding windows is that capacity increases in the same pattern as the usage in the previous window. E.g., If you get burst of 10 requests, an idle interval and then another burst that exausts your rate limit, in the next window you will also get an increase of 10 in the burst capacity, then no changes during an interval and then full burst capacity. Instead it is preferable to increase it continuously every `window/burst` seconds (see Leaky Bucket below).
```

```python
class SlidingWindowRateLimiter:
    def __init__(self, burst: int, window: float):
        self.burst = burst
        self.window = window
        self.request_timestamps = []  # timestamps of requests in the current window (sorted)

    def __call__(self) -> bool:
        now = time.time()
        window_start = now - self.window

        # remove request timestamps that are now outside the sliding window
        while self.request_timestamps and self.request_timestamps[0] < window_start:
            self.request_timestamps.pop(0)

        if len(self.request_timestamps) >= self.burst:
            # reject the request if the current window is exausted
            return False
        else:
            # insert the current timestamp in the window
            self.request_timestamps.append(now)
            return True
```

## GCRA (aka Leaky Bucket)

The algorithm we use, is awesome.

> TODO

Reviewer: I appreciate the rundown of the other algorithms but this is probably the most important one to explain, since it's the one the library actually uses - is the plan to add this before merge @paulo-raca?
